serif_cli/src/main/kotlin/serif_cli/App.kt

```diff
@@ -2,15 +2,57 @@
  * This Kotlin source file was generated by the Gradle 'init' task.
  */
 package xyz.room409.serif.serif_cli
-import xyz.room409.serif.serif_shared.Greeting
+import xyz.room409.serif.serif_shared.*
+
+import java.io.Console
 
 class App {
-    val greeting: String
+    val console = System.console()
+    var mstate: MatrixState = MatrixLogin()
+    val version: String
         get() {
-            return "from cli: " + Greeting().greeting()
+            return mstate.version + ", CLI UI"
+        }
+    fun run() {
+        if (console == null) {
+            println()
+            println("Console was null! Probs running under gradle.")
+            println("It doesn't work, and is really irritating")
+            println("Try ./run_dist.sh if you're on a Unixy OS")
+            println("Otherwise, if you're on Windows, you'll have to manually")
+            println("do ./gradlew :serif_cli:assembleDist and then find and")
+            println("extract the distribution zip, and then run the batch file")
+            println("in the bin folder. Probs best to automate the process")
+            println("like run_dist.sh, but I'm not on Windows and can't")
+            println("a batch script without a reference. My apologies.")
+            println("-Nathan")
+            println()
+            return
+        }
+        while (true) {
+            val m = mstate
+            mstate = when (m) {
```

Comment:
Good question! Kotlin's smart cast only works on immutable variables, because otherwise the variable could have changed value between the "when" and the use of the value. Temporarily assigning the value to an immutable m variable allows the smart cast to work.